src/main/java/apoc/nodes/Nodes.java

```diff
@@ -136,6 +139,94 @@ public boolean connected(@Name("start") Node start, @Name("start") Node end, @Na
+        if (first.getRelationships().iterator().hasNext()) {
+            return StreamSupport.stream(first.getRelationships().spliterator(), false)
+                    .map(relationship -> new VirtualPathResult(relationship.getStartNode(), relationship, relationship.getEndNode()));
+        } else {
+            return Stream.of(new VirtualPathResult(first, null, null));
+        }
+    }
+
+    private VirtualNode createVirtualNode(Set<Node> nodes, RefactorConfig conf) {
+        Create create = new Create();
+        Node first = nodes.iterator().next();
+        List<String> labels = Util.labelStrings(first);
+        if (conf.isCollapsedLabel()) {
+            labels.add("Collapsed");
+        }
+        VirtualNode virtualNode = (VirtualNode) create.vNodeFunction(labels, first.getAllProperties());
+        createVirtualRelationships(nodes, virtualNode, first, conf);
+        nodes.stream().skip(1).forEach(node -> {
+            node.getLabels().forEach(label -> { // Set all Label
+                if (!virtualNode.hasLabel(label)) {
+                    virtualNode.addLabel(label);
+                }
+            });
+            mergePropertiesWithCount(node.getAllProperties(), virtualNode, conf); // Set all properties
+            createVirtualRelationships(nodes, virtualNode, node, conf);
+        });
+        if (conf.isCountMerge()) {
+            virtualNode.setProperty("countNodes", nodes.size()); // Count nodes merged
+        }
+        return virtualNode;
+    }
+
+    private void createVirtualRelationships(Set<Node> nodes, VirtualNode virtualNode, Node node, RefactorConfig refactorConfig) {
+        node.getRelationships().forEach(relationship -> {
+            Node startNode = relationship.getStartNode();
+            Node endNode = relationship.getEndNode();
+
+            if (nodes.contains(startNode) && nodes.contains(endNode)) {
+                if (refactorConfig.isSelfRel()) {
+                    createOrMergeVirtualRelationship(virtualNode, refactorConfig, relationship, virtualNode,  Direction.OUTGOING);
+                }
+            } else {
+                if (startNode.getId() == node.getId()) {
+                    createOrMergeVirtualRelationship(virtualNode, refactorConfig, relationship, endNode,  Direction.OUTGOING);
+                } else {
+                    createOrMergeVirtualRelationship(virtualNode, refactorConfig, relationship, startNode,  Direction.INCOMING);
+                }
+            }
+        });
+    }
+
+    private void createOrMergeVirtualRelationship(VirtualNode virtualNode, RefactorConfig refactorConfig, Relationship source, Node node, Direction direction) {
+        Iterable<Relationship> iterator = virtualNode.getRelationships(source.getType(), direction);
+        Optional<Relationship> first = StreamSupport.stream(iterator.spliterator(), false).filter(relationship -> relationship.getEndNode().equals(node)).findFirst();
```

Comment:
getEndNode() is wrong for direction INCOMING (or BOTH) you need to use rel.getOtherNode(virtualNode)
also make sure there is a test for both incoming and outgoing rels